### PR TITLE
BF16 cutedsl prune [2] + cache bf16 cutedsl kernels

### PR DESCRIPTION
**Pull request by jiahanc**
## 📌 Description

Follow-up to https://github.com/flashinfer-ai/flashinfer/pull/5089.

Add persistent compiled-kernel caching for `mm_bf16(backend="cute-dsl")` and conservatively prune cluster Split-K stages.

- Reuse `build_and_load_cute_dsl_kernel()` for Direct, cluster Split-K, and warp Split-K. Preserve specialization fields, device selection, and the current PyTorch stream through TVM-FFI.
- Cap cluster stage candidates at a non-wrapping power-of-two ring derived from each CTA's K-tile count. Kernel mainloops, default tactics, and Direct/warp candidate sets are unchanged.
- Remove the manually bumped BF16 autotune cache-version field; retain dtype, output dtype, bias, PDL, and architecture extras.

### Measurements

B300 SXM6 AC, Torch 2.13.0+cu130, CuTeDSL 4.7.0; BF16 input/output, no bias, PDL enabled; default M buckets 1/2/4/8/16/32.

The disk-cache experiment below fully retuned in **separate processes**, without persisting autotune winners. Common binding setup is excluded. These cache measurements preceded the additional stage pruning.

| N | K | Empty disk cache: JIT + export + autotune | New process: disk load + autotune | CuTe compilations on reload |
|---:|---:|---:|---:|---:|
| 4608 | 8192 | 563.784 s | 8.021 s | 0 |
| 8192 | 2048 | 351.925 s | 6.930 s | 0 |

First use still pays JIT/export costs; this is **not** an offline-precompiled package or a claim of faster first-ever compilation.

Cluster candidates per M bucket:

| K | Before | After |
|---:|---:|---:|
| 128 | 35 | 12 |
| 256 | 58 | 24 |
| 512 | 85 | 53 |
| 1024 | 41 | 29 |
| 2048 | 41 | 37 |
| 8192 | 41 | 41 |

Hot/cold CUPTI candidate-set comparisons covered all 32 shapes in #4266 affected by this pruning; no performance regression was observed. The other 32 shapes retain identical candidate sets. This measures the best available candidate latency, not a guarantee about a noisy production autotuner's selection.

<details>
<summary>Reproduce cache reuse</summary>

Run this program twice in independent Python processes with the same `FLASHINFER_WORKSPACE_BASE`, initially using an empty task-specific cache directory. Change `n,k` for the second matrix family. Do not save/load an autotune winner cache.

```python
import time
import torch
import flashinfer
from flashinfer.gemm import gemm_base as gb

n, k = 4608, 8192
a = torch.randn((32, k), device="cuda", dtype=torch.bfloat16)
b = torch.randn((n, k), device="cuda", dtype=torch.bfloat16).T
out = torch.empty((32, n), device="cuda", dtype=torch.bfloat16)
workspace = gb._get_cache_buf("mm_bf16_workspace", gb.DEFAULT_WORKSPACE_SIZE, a.device)
gb._cute_dsl_bf16_runners([a, b, None, True, out, workspace])
torch.cuda.synchronize()
start = time.perf_counter()
with flashinfer.autotune(True):
    flashinfer.mm_bf16(a, b, out=out, pdl=True, backend="cute-dsl")
torch.cuda.synchronize()
print(time.perf_counter() - start)
```

Only compiled artifacts are persisted by this change; each process still performs autotuning.

</details>

## 🔍 Related Issues

Related: #4266, #5089.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

No test files are added or modified; existing accuracy tests are reused.

- Post-rebase B300: `pytest -q tests/gemm/test_mm_bf16.py -k 'cute and not cutile'` — **10 passed** on CuTeDSL 4.7.0 and **10 passed** on 4.6.2.
- Fresh-process cache reload verified with CuTe compilation forbidden; eager, custom-stream, CUDA Graph, bias, BF16/FP16, and cache-disabled paths checked locally.
- Additional local numerical checks: 178 cache-path cases, 256 stage-boundary cases, and actual M=1..32 after autotuning both matrix families.
- Cache integration: 19 matched-tactic cases across hot/cold L2, maximum observed slowdown 0.55%. Stage pruning: broader 54-cell check within a 1% near-tie band; the final 20-shape/40-cell #4266 follow-up measured changes of -0.24% to 0%.
- `pre-commit run --all-files` passed on the final rebased revision.

## 🔬 Experimental Track

- [ ] This PR is **experimental**: it adds or changes code under `flashinfer/experimental/` and/or an `@flashinfer_experimental_api`. Tracking issue: #
  - [ ] The tracking issue names an owner, the reason for the experimental path, and a graduation plan with a target release.
  - [ ] Core changes are limited to a thin entry point (signature, shared validation, feature-gate check, backend selection, handoff).
  - [ ] Tests live in `tests/experimental/` and were validated on the intended hardware; a runnable example is included.
  - [ ] Nothing is registered in `flashinfer/aot.py`, and no experimental backend is reachable from `backend="auto"` without `FLASHINFER_ALLOW_EXPERIMENTAL_AUTO_BACKENDS=1`. (Calling an `@flashinfer_experimental_api` or naming a backend explicitly is itself the opt-in and needs no environment variable.)
  - [ ] **Test scope declared below.** The experimental CI lane runs exactly these targets, so keep them as narrow as the change allows.

```experimental-tests
# One target per line: a directory or a file. (A pytest ::selector is not
# supported -- the sharding runner cannot consume one.) Must be under
# tests/experimental/ and must exist. Delete these comment lines and add yours, e.g.
#
#   tests/experimental/test_my_backend.py
#   tests/experimental/my_backend/
#
# Declaring the whole tree (tests/experimental/) is allowed but means every
# experimental PR pays for every other feature's tests, in every matrix cell.
```

## Reviewer Notes

The diff is limited to the three existing kernel modules and the BF16 interface documentation/cache extras. Local benchmarks, result files, and logs are not included.

Dynamic M, offline wheel generation, and the rejected B-loader pruning experiment are out of scope.

## Summary by CodeRabbit

* **New Features**
  * Added disk caching for compiled BF16 GEMM kernels, enabling reuse across processes and application runs.
  * Improved cache reuse across compatible GEMM shapes and configurations.
  * Preserved the caller’s PyTorch CUDA stream during kernel execution.
  * Added deployment guidance for retaining or configuring the kernel cache location.

* **Documentation**
  * Documented BF16 GEMM disk-cache behavior and deployment considerations.

### CODE REVIEW

**Review by StudyingShao — APPROVED** [review]
LGTM.
